### PR TITLE
[AD-329] Refactor to use $project when possible

### PR DESCRIPTION
**Pull request by andiemontoyeah**
### Summary

- Refactor DocumentDbProject and DocumentDbFilter to use $project when possible.  

### Description

- Added logic to check if current subtree is a part of a "join" to determine whether it is ok to use $project at that time. 
- Removed rule to "optimize" by removing project in SELECT * queries since this is actually not desired. :)
- Updated unit tests -> moved filter tests to own separate file since current file exceeded 2000 lines 

### Related Issue

https://bitquill.atlassian.net/browse/AD-329

### Additional Reviewers
@birschick-bq
@mitchell-elholm
@alexey-temnikov 
@andiem-bq
@affonsoBQ

### CODE REVIEW

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbProject.java [hunk 1]
```diff
@@ -122,12 +122,29 @@ public DocumentDbProject(final RelOptCluster cluster, final RelTraitSet traitSet
...
+                            .sqlName(oldColumn.getSqlName())
+                            .sqlType(oldColumn.getSqlType())
+                            .dbType(oldColumn.getDbType())
+                            .isIndex(oldColumn.isIndex())
+                            .isPrimaryKey(oldColumn.isPrimaryKey())
+                            .foreignKeyTableName(oldColumn.getForeignKeyTableName())
+                            .foreignKeyColumnName(oldColumn.getForeignKeyColumnName())
+                            .resolvedPath(outName)
+                            .build();
                     columnMap.remove(inName);
```
Comment (birschick-bq) [inline calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbProject.java]: Do we need this line, now? Looks like you already removed this in line 127.

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbProject.java [hunk 2]
```diff
@@ -122,12 +122,29 @@ public DocumentDbProject(final RelOptCluster cluster, final RelTraitSet traitSet
...
+                    columnMap.put(outName, newColumn);
```
Comment (birschick-bq) [inline calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbProject.java:146]: Do we need to maintain "in-order" column map? Will this cause us a problem?

File: calcite-adapter/src/test/java/software/amazon/documentdb/jdbc/query/DocumentDbQueryMappingServiceLiteralTest.java [hunk 3]
```diff
@@ -215,17 +256,24 @@ void testQueryForObjectId() throws SQLException {
...
-                result6.getAggregateOperations().get(1));
+                "{\"$match\": {" + DocumentDbFilter.BOOLEAN_FLAG_FIELD + ": {\"$eq\": true}}}"),
+                result1.getAggregateOperations().get(1));
+        Assertions.assertEquals(BsonDocument.parse(
+                "{\"$project\": {" + DocumentDbFilter.BOOLEAN_FLAG_FIELD + ": 0}}"),
+                result1.getAggregateOperations().get(2));
         Assertions.assertEquals(BsonDocument.parse(
-                        "{\"$project\": {\"placeholderField1F84EB1G3K47\": 0}}"),
-                result6.getAggregateOperations().get(2));
+                "{\"$project\": {\"objectIdCollection__id\": \"$_id\", \"_id\": 0}}"),
```
Comment (birschick-bq) [inline calcite-adapter/src/test/java/software/amazon/documentdb/jdbc/query/DocumentDbQueryMappingServiceLiteralTest.java:276]: Ah. OK. I get it. We are projecting with the mapped name - so we don't need the _id column, too.

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbProject.java (hunk 2 above)
Comment (andiemontoyeah) [reply calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbProject.java:146]: i dont think order matters but the lack of order did affect some unit tests where the $exists filters appear in a different order (different json string but same meaning) depending on the query

File: calcite-adapter/src/test/java/software/amazon/documentdb/jdbc/query/DocumentDbQueryMappingServiceLiteralTest.java (hunk 3 above)
Comment (andiemontoyeah) [reply calcite-adapter/src/test/java/software/amazon/documentdb/jdbc/query/DocumentDbQueryMappingServiceLiteralTest.java:276]: yeah! 😅  we aren't sure why yet but the slightly smaller output document greatly improves performance for some cases

**Review by birschick-bq — APPROVED** [review]

**Pull request merged**